src/download.rs

```diff
@@ -71,6 +82,47 @@ pub fn download(config: &DownloadConfig) -> Result<()> {
     Ok(())
 }
 
+// Check an existing image and signature for validity.  The image cannot
+// have been decompressed after downloading.  Return an error if invalid for
+// any reason.
+fn check_image_and_sig(source: &ImageSource, path: &Path, sig_path: &Path) -> Result<()> {
+    // ensure we have something to check
+    if source.signature.is_none() {
+        return Err("no signature available; can't check existing file".into());
+    }
+    let signature = source.signature.as_ref().unwrap();
+
+    // compare signature to expectation
+    let mut sig_file = OpenOptions::new()
+        .read(true)
+        .open(sig_path)
+        .chain_err(|| format!("opening {}", sig_path.display()))?;
+    let mut buf = Vec::new();
+    sig_file
+        .read_to_end(&mut buf)
+        .chain_err(|| format!("reading {}", sig_path.display()))?;
+    if &buf != signature {
+        return Err("signature file doesn't match source".into());
+    }
+
+    // open image file
+    let mut file = OpenOptions::new()
+        .read(true)
+        .open(path)
+        .chain_err(|| format!("opening {}", path.display()))?;
+
+    // perform GPG verification
+    let mut verify_reader = GpgReader::new(&mut file, signature)?;
+    let mut buf: [u8; 4096] = [0; 4096];
+    while verify_reader
+        .read(&mut buf)
+        .chain_err(|| "reading image file")?
+        > 0
```

Comment:
It looks like this logic may be better as its own method attached to `GpgReader`.